File: go/common/lru_cache_test.go
// Copyright (c) 2024 Fantom Foundation
//
// Use of this software is governed by the Business Source License included
// in the LICENSE file and at fantom.foundation/bsl11.
//
// Change Date: 2028-4-16
//
// On the date above, in accordance with the Business Source License, use of
// this software will be governed by the GNU Lesser General Public License v3.

package common

import (
	"testing"
)

func TestLruExceedCapacity(t *testing.T) {
	c := NewLruCache[int, int](3)

	c.Set(1, 11)
	c.Set(2, 22)

	evictedKey, evictedValue, evicted := c.Set(3, 33)
	if evictedKey != 0 || evictedValue != 0 || evicted {
		t.Errorf("No items should have been evicted yet")
	}

	_, exists := c.Get(1) // one refreshed - first in the list now
	if exists == false {
		t.Errorf("Item should exist")
	}

	evictedKey, evictedValue, evicted = c.Set(5, 44)
	if evictedKey != 2 || evictedValue != 22 || evicted == false {
		t.Errorf("Incorrectly evicted items: %d/%d", evictedKey, evictedValue)
	}
	_, exists = c.Get(2) // 2 is the oldest in the table
	if exists {
		t.Errorf("Item should be evicted")
	}
}

// TestLRUOrder test correct ordering of the keys
func TestLRUOrder(t *testing.T) {
	c := NewLruCache[int, int](3)

	c.Set(1, 11)
	c.Set(2, 22)
	c.Set(3, 33)

	_, _ = c.Get(1) // one refreshed - first in the list now
	if c.head.key != 1 {
		t.Errorf("Item should be head")
	}
	if c.tail.key != 2 {
		t.Errorf("Item should be tail")
	}

	c.Set(2, 222) // two refreshed - first in the list now
	if c.head.key != 2 {
		t.Errorf("Item should be head")
	}
	if c.tail.key != 3 {
		t.Errorf("Item should be tail")
	}

	// insert exceeding and check order
	c.Set(4, 44)
	if c.head.key != 4 || c.head.next.key != 2 || c.head.next.next.key != 1 {
		t.Errorf("wrong order")
	}
	if c.tail.key != 1 || c.tail.prev.key != 2 || c.tail.prev.prev.key != 4 {
		t.Errorf("wrong order")
	}
}

func TestLRUCache_GetOrSet(t *testing.T) {
	c := NewLruCache[int, int](4)

	if _, present, _, _, evicted := c.GetOrSet(1, 11); present || evicted {
		t.Errorf("value should be neither present nor evicted")
	}

	if current, present, _, _, _ := c.GetOrSet(1, 12); !present || current != 11 {
		t.Errorf("previous value should be present")
	}

	// cause eviction
	c.Set(5, 5)
	c.Set(9, 9)
	c.Set(13, 13)

	if _, present, evictedKey, evictedValue, evicted := c.GetOrSet(17, 13); !evicted || present || evictedKey != 1 || evictedValue != 11 {
		t.Errorf("value should be evicted: %d != 1 || %d != 11", evictedKey, evictedValue)
	}

	// no eviction - replacing
	if current, present, _, _, evicted := c.GetOrSet(9, 13); evicted || !present || current != 9 {
		t.Errorf("value should be evicted: %d != 9", current)
	}

}

func TestCache_Entry_String(t *testing.T) {
	e := entry[int, int]{10, 20, nil, nil}

	if got, want := e.String(), "Entry: 10 -> 20"; got != want {
		t.Errorf("provided string does not match: %s != %s", got, want)
	}
}

File: cpp/backend/common/cache/lru_cache_test.cc
/*
 * Copyright (c) 2024 Fantom Foundation
 *
 * Use of this software is governed by the Business Source License included
 * in the LICENSE.TXT file and at fantom.foundation/bsl11.
 *
 * Change Date: 2028-4-16
 *
 * On the date above, in accordance with the Business Source License, use
 * of this software will be governed by the GNU Lesser General Public Licence v3.
 */

#include "backend/common/cache/lru_cache.h"

#include <type_traits>

#include "gmock/gmock.h"
#include "gtest/gtest.h"

namespace carmen::backend::index {
namespace {

using ::testing::ElementsAre;
using ::testing::IsNull;
using ::testing::Pointee;

using TestCache = LeastRecentlyUsedCache<int, int>;

TEST(LeastRecentlyUsedCacheTest, MissingElementsReturnNullptr) {
  TestCache cache(2);
  EXPECT_THAT(cache.Get(0), IsNull());
  EXPECT_THAT(cache.Get(1), IsNull());
}

TEST(LeastRecentlyUsedCacheTest, ElementsCanBeSetAndRetrieved) {
  TestCache cache(2);
  cache.Set(0, 1);
  EXPECT_THAT(cache.Get(0), Pointee(1));
}

TEST(LeastRecentlyUsedCacheTest, KeysAreDifferentiated) {
  TestCache cache(2);
  cache.Set(0, 1);
  cache.Set(1, 2);
  EXPECT_THAT(cache.Get(0), Pointee(1));
  EXPECT_THAT(cache.Get(1), Pointee(2));
}

TEST(LeastRecentlyUsedCacheTest, KeysCanBeUpdated) {
  TestCache cache(2);
  EXPECT_THAT(cache.Get(0), IsNull());
  cache.Set(0, 1);
  EXPECT_THAT(cache.Get(0), Pointee(1));
  cache.Set(0, 2);
  EXPECT_THAT(cache.Get(0), Pointee(2));
}

TEST(LeastRecentlyUsedCacheTest, SizeLimitIsEnforced) {
  TestCache cache(2);
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre());
  cache.Set(0, 1);
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre(0));
  cache.Set(1, 2);
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre(1, 0));
  cache.Set(2, 3);
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre(2, 1));
  cache.Set(0, 4);
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre(0, 2));
}

TEST(LeastRecentlyUsedCacheTest, GetIsATouch) {
  TestCache cache(3);
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre());
  cache.Set(1, 1);
  cache.Set(2, 2);
  cache.Set(3, 3);
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre(3, 2, 1));

  // Get missing.
  EXPECT_EQ(cache.Get(0), nullptr);
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre(3, 2, 1));

  // Get first in LRU order.
  EXPECT_THAT(cache.Get(3), Pointee(3));
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre(3, 2, 1));

  // Get middle in LRU order.
  EXPECT_THAT(cache.Get(2), Pointee(2));
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre(2, 3, 1));

  // Get last in LRU order.
  EXPECT_THAT(cache.Get(1), Pointee(1));
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre(1, 2, 3));
}

TEST(LeastRecentlyUsedCacheTest, SetIsATouch) {
  TestCache cache(3);
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre());
  cache.Set(1, 1);
  cache.Set(2, 2);
  cache.Set(3, 3);
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre(3, 2, 1));

  // Set first in LRU order.
  cache.Set(3, 0);
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre(3, 2, 1));

  // Get middle in LRU order.
  cache.Set(2, 0);
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre(2, 3, 1));

  // Get last in LRU order.
  cache.Set(1, 0);
  EXPECT_THAT(cache.GetOrderedKeysForTesting(), ElementsAre(1, 2, 3));
}

}  // namespace
}  // namespace carmen::backend::index
